lib/cursor/QueryCursor.js

```diff
@@ -295,40 +295,102 @@ function _next(ctx, cb) {
   }
 
   if (ctx.cursor) {
-    return ctx.cursor.next(function(error, doc) {
-      if (error) {
-        return callback(error);
-      }
-      if (!doc) {
+    if (ctx.query._mongooseOptions.populate && !ctx._pop) {
+      ctx._pop = helpers.preparePopulationOptionsMQ(ctx.query,
+        ctx.query._mongooseOptions);
+      ctx._pop.__noPromise = true;
+    }
+    if (ctx.query._mongooseOptions.populate && ctx.options.batchSize > 1) {
+      if (ctx._batchDocs && ctx._batchDocs.length) {
+        // Return a cached populated doc
+        return _nextDoc(ctx, ctx._batchDocs.shift(), ctx._pop, callback);
+      } else if (ctx._batchExhausted) {
+        // Internal cursor reported no more docs. Act the same here
         return callback(null, null);
+      } else {
+        // Request as many docs as batchSize, to populate them also in batch
+        return ctx.cursor.next(_onNext.bind({ ctx, callback, batchDocs: [] }));
```

Comment:
I'd prefer we not use `bind()`, especially with a custom object as `this`.

Reply:
Yeah, that was a quick approach to make ESLint happy (you can see the previous commit). Take in mind `callback` needs to be preserved through the internal cursor iteration. There probably could be a cleaner way to succeed in this. Any suggestions?

I also did not wanted to expose `onNext`, `onPopulate` and `nextDoc` in the QueryCursor API, as they are purely internal methods. That forces me to pass somehow the `ctx` object in any way. And thus i ended with the `bind()` approach.